plugins/accounts-settings.c

```diff
@@ -0,0 +1,474 @@
+{
+	DBusConnection *conn = ofono_dbus_get_connection();
+	size_t i;
+
+	snprintf(accounts->uid_path, sizeof(accounts->uid_path),
+			ACCOUNTS_PATH"%u", (unsigned) accounts->current_uid);
+
+	/*
+	 * Register for property changes. Note that AccountsService is D-Bus,
+	 * initiated, so there is no risk for race conditions on start.
+	 */
+	accounts->prop_change_watch =
+		g_dbus_add_signal_watch(conn, ACCOUNTS_SERVICE,
+						accounts->uid_path,
+						DBUS_INTERFACE_PROPERTIES,
+						"PropertiesChanged",
+						property_changed,
+						accounts, NULL);
+
+	/* Retrieve AccountService properties */
+	for (i = 0; i < NUM_SETTINGS; ++i)
+		get_property(accounts, i);
+}
+
+static void release_accounts_data(struct accounts_data *accounts)
+{
+	DBusConnection *conn = ofono_dbus_get_connection();
+	size_t i;
+
+	g_dbus_remove_watch(conn, accounts->prop_change_watch);
+
+	for (i = 0; i < NUM_SETTINGS; ++i) {
+		g_free(accounts->property_values[i]);
+		accounts->property_values[i] = NULL;
+	}
+}
+
+static uid_t get_active_seat_uid(void)
+{
+	char **seats, **iter;
+	int res;
+	gboolean found = FALSE;
+	uid_t uid;
+
+	res = sd_get_seats(&seats);
+	if (res < 0) {
+		ofono_error("Error retrieving seats: %s (%d)",
+							strerror(-res), -res);
+		goto end;
+	} else if (res == 0) {
+		ofono_info("No seats found");
+		goto end;
+	}
+
+	for (iter = seats; *iter; ++iter) {
+
+		if (!found && sd_seat_get_active(*iter, NULL, &uid) >= 0) {
+			DBG("seat %s with uid %d", *iter, uid);
+			found = TRUE;
+		}
+
+		free(*iter);
+	}
+
+	free(seats);
+
+end:
+	if (!found)
+		uid = geteuid();
+
+	return uid;
+}
+
+static gboolean sd_changed(GIOChannel *stream, GIOCondition condition,
+							gpointer user_data)
+{
+	struct accounts_data *accounts = user_data;
+	uid_t new_uid;
+
+	sd_login_monitor_flush(accounts->login_monitor);
+
+	new_uid = get_active_seat_uid();
+	if (new_uid != accounts->current_uid) {
+		/* User in seat changed, resetting data */
+		release_accounts_data(accounts);
+
+		accounts->current_uid = new_uid;
+		get_user_settings(accounts);
+	}
+
+	return TRUE;
+}
+
+static void sd_init(struct accounts_data *accounts)
+{
+	int status;
+	GIOChannel *stream;
+	int fd;
+
+	status = sd_login_monitor_new(NULL, &accounts->login_monitor);
+	if (status < 0) {
+		ofono_error("Error creating systemd login monitor: %d", status);
+		accounts->login_monitor = NULL;
+		return;
+	}
+
+	fd = sd_login_monitor_get_fd(accounts->login_monitor);
+	stream = g_io_channel_unix_new(fd);
+	accounts->login_watch =
+			g_io_add_watch(stream, G_IO_IN, sd_changed, accounts);
+
+	g_io_channel_unref(stream);
+}
+
+static void sd_finalize(struct accounts_data *accounts)
+{
+	if (accounts->login_monitor == NULL)
+		return;
+
+	g_source_remove(accounts->login_watch);
+	sd_login_monitor_unref(accounts->login_monitor);
+}
+
+static int accounts_settings_init(void)
+{
+	accounts_g.current_uid = get_active_seat_uid();
+
+	/* Register for login changes */
+	sd_init(&accounts_g);
```

Comment:
Why do you pass the address of a static variable to this function?  Wouldn't it be simpler to just use the static variable directly in all cases?